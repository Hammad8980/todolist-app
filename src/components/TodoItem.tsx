export type Task = {
  id: number;
  title: string;
  isCompleted: boolean;
  priority?: 'p1' | 'p2' | 'p3';
};
