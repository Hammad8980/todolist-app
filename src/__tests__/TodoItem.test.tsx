import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoItem from '../components/todo/TodoItem';
import type { Task } from '../features/todos/TodoTaskTypes';

// Mock the Button component
jest.mock('../components/ui/Button', () => {
  return function MockButton({
    children,
    onClick,
    className,
    ...props
  }: {
    children: React.ReactNode;
    onClick: () => void;
    className?: string;
    [key: string]: unknown; //Allows any additional props (disabled, aria-label, type, etc.)
  }) {
    return (
      <button onClick={onClick} className={className} {...props}>
        {children}
      </button>
    );
  };
});

describe('TodoItem', () => {
  const mockOnDelete = jest.fn();
  const mockOnToggle = jest.fn();

  const sampleTask: Task = {
    id: 1,
    title: 'Sample task',
    isCompleted: false,
    priority: 'p1',
  };

  beforeEach(() => {
    mockOnDelete.mockClear();
    mockOnToggle.mockClear();
  });

  it('renders task title and checkbox correctly', () => {
    const { container } = render(
      <ul>
        <TodoItem task={sampleTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    expect(screen.getByText('Sample task')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).not.toBeChecked();
    expect(screen.getByRole('button', { name: 'Delete' })).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders completed task with correct styles', () => {
    const completedTask: Task = {
      ...sampleTask,
      isCompleted: true,
    };

    const { container } = render(
      <ul>
        <TodoItem task={completedTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    const taskText = screen.getByText('Sample task');
    expect(taskText).toHaveClass('line-through', 'text-gray-500');
    expect(screen.getByRole('checkbox')).toBeChecked();
    expect(container).toMatchSnapshot();
  });

  it('calls onToggle when checkbox is clicked', async () => {
    const user = userEvent.setup();
    const { container } = render(
      <ul>
        <TodoItem task={sampleTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    const checkbox = screen.getByRole('checkbox');
    await user.click(checkbox);

    expect(mockOnToggle).toHaveBeenCalledWith(1);
    expect(mockOnToggle).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it('calls onDelete when delete button is clicked', async () => {
    const user = userEvent.setup();
    const { container } = render(
      <ul>
        <TodoItem task={sampleTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    const deleteButton = screen.getByRole('button', { name: 'Delete' });
    await user.click(deleteButton);

    expect(mockOnDelete).toHaveBeenCalledWith(1);
    expect(mockOnDelete).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it('handles checkbox change via fireEvent', () => {
    const { container } = render(
      <ul>
        <TodoItem task={sampleTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    const checkbox = screen.getByRole('checkbox');
    fireEvent.click(checkbox);

    expect(mockOnToggle).toHaveBeenCalledWith(1);
    expect(mockOnToggle).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it('renders children when provided', () => {
    const { container } = render(
      <ul>
        <TodoItem task={sampleTask} onDelete={mockOnDelete} onToggle={mockOnToggle}>
          <span>Additional content</span>
        </TodoItem>
      </ul>
    );

    expect(screen.getByText('Additional content')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('handles different priority values', () => {
    const highPriorityTask: Task = {
      ...sampleTask,
      priority: 'p1',
    };

    const { container } = render(
      <ul>
        <TodoItem task={highPriorityTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    expect(screen.getByText('Sample task')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('handles task without priority', () => {
    const taskWithoutPriority: Task = {
      id: 2,
      title: 'Task without priority',
      isCompleted: false,
    };

    const { container } = render(
      <ul>
        <TodoItem task={taskWithoutPriority} onDelete={mockOnDelete} onToggle={mockOnToggle} />
      </ul>
    );

    expect(screen.getByText('Task without priority')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('handles multiple tasks with different states', () => {
    const taskNotCompleted: Task = { id: 1, title: "Task 1", isCompleted: false, priority: 'p1' };
    const taskCompleted: Task = { id: 2, title: "Task 2", isCompleted: true, priority: 'p3' };
    const taskWithoutPriority: Task = { id: 3, title: "Task 3", isCompleted: false };

    const { rerender, container } = 
    render(<TodoItem task={taskNotCompleted} onDelete={mockOnDelete} onToggle={mockOnToggle} />);
    expect(screen.getByText('Task 1')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).not.toBeChecked();
    expect(container).toMatchSnapshot('taskNotCompleted');

    rerender(<TodoItem task={taskCompleted} onDelete={mockOnDelete} onToggle={mockOnToggle} />);
    expect(screen.getByText('Task 2')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).toBeChecked();
    expect(container).toMatchSnapshot('taskCompleted');

    rerender(<TodoItem task={taskWithoutPriority} onDelete={mockOnDelete} onToggle={mockOnToggle} />);
    expect(screen.getByText('Task 3')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).not.toBeChecked();
    expect(container).toMatchSnapshot('taskWithoutPriority');
  });

  it('does not call callbacks on render', () => {
    const { container } = render(<TodoItem task={sampleTask} onDelete={mockOnDelete} onToggle={mockOnToggle} />);

    expect(mockOnDelete).not.toHaveBeenCalled();
    expect(mockOnToggle).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
