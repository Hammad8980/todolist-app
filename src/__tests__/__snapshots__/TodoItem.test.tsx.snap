// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem calls onDelete when delete button is clicked 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoItem calls onToggle when checkbox is clicked 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoItem does not call callbacks on render 1`] = `
<div>
  <li
    class="flex items-center gap-2 py-2 border-b"
  >
    <input
      class="h-5 w-5"
      type="checkbox"
    />
    <span
      class=""
    >
      Sample task
    </span>
    <button
      class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
    >
      <img
        alt="Delete"
        class="w-6 h-6"
        src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
        style="filter: invert(1) brightness(2);"
      />
    </button>
  </li>
</div>
`;

exports[`TodoItem handles checkbox change via fireEvent 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoItem handles different priority values 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoItem handles multiple tasks with different states: taskCompleted 1`] = `
<div>
  <li
    class="flex items-center gap-2 py-2 border-b"
  >
    <input
      class="h-5 w-5"
      type="checkbox"
    />
    <span
      class="line-through text-gray-500"
    >
      Task 2
    </span>
    <button
      class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
    >
      <img
        alt="Delete"
        class="w-6 h-6"
        src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
        style="filter: invert(1) brightness(2);"
      />
    </button>
  </li>
</div>
`;

exports[`TodoItem handles multiple tasks with different states: taskNotCompleted 1`] = `
<div>
  <li
    class="flex items-center gap-2 py-2 border-b"
  >
    <input
      class="h-5 w-5"
      type="checkbox"
    />
    <span
      class=""
    >
      Task 1
    </span>
    <button
      class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
    >
      <img
        alt="Delete"
        class="w-6 h-6"
        src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
        style="filter: invert(1) brightness(2);"
      />
    </button>
  </li>
</div>
`;

exports[`TodoItem handles multiple tasks with different states: taskWithoutPriority 1`] = `
<div>
  <li
    class="flex items-center gap-2 py-2 border-b"
  >
    <input
      class="h-5 w-5"
      type="checkbox"
    />
    <span
      class=""
    >
      Task 3
    </span>
    <button
      class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
    >
      <img
        alt="Delete"
        class="w-6 h-6"
        src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
        style="filter: invert(1) brightness(2);"
      />
    </button>
  </li>
</div>
`;

exports[`TodoItem handles task without priority 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Task without priority
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoItem renders children when provided 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
      <span>
        Additional content
      </span>
    </li>
  </ul>
</div>
`;

exports[`TodoItem renders completed task with correct styles 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        checked=""
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class="line-through text-gray-500"
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;

exports[`TodoItem renders task title and checkbox correctly 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 py-2 border-b"
    >
      <input
        class="h-5 w-5"
        type="checkbox"
      />
      <span
        class=""
      >
        Sample task
      </span>
      <button
        class="ml-auto bg-red-500 hover:bg-red-700 text-white text-sm py-1 px-2 rounded"
      >
        <img
          alt="Delete"
          class="w-6 h-6"
          src="https://img.icons8.com/?size=50&id=OD5jprZTbcDK&format=png"
          style="filter: invert(1) brightness(2);"
        />
      </button>
    </li>
  </ul>
</div>
`;
